### PR TITLE
CytoSnake New Pathing implementation

### PR DESCRIPTION
**Pull request by axiomcura**
- [About](#about)
- [Motivation](#motivation)
- [Implementation Approach](#implementation-approach)
- [Reviewer focus](#reviewer-focus)
  - [Greg](#greg)
  - [Dave](#dave)
- [Usage and assumption](#usage-and-assumption)
  - [Init mode](#init-mode)
  - [Run mode](#run-mode)
  - [Help mode](#help-mode)

## About

This PR introduces a new pathing system for `CytoSnake`. This PR is sections into multiple parts:

- Motivation
- Implementation Approach
- Reviewer Focus
- Usage and assumption

The **Motivation** explains the whole purpose of this PR. Explains the issues there we encountered with the previous version and provide justification why this PR is necessary

The **implementation approach** provides a detailed description on how the issue was resolved. This will describe the types of software engineering approaches were conducted in order to solve this issue.

In the **reviewer focuses** section, reaches out to main reviewers on what to focus on. This allows reviewers were to focus on instead of trying to find on "what to do".

In the **usage and assumptions** sections provides a description on how `cytosnake` is used. Detailed explanations will be provided including what are the assumptions present when using `cytosnake`.

## Motivation

CytoSnake is a CLI tool that contains multiple reproducible workflows that analyzes cell morphology readouts.

Recently, `CytoSnake` has gone through it's first round of usage testing by @jenna-tomkinson  and pointed out some major issues in regards of `CytoSnake's` strict naming scheme.

`CytoSnake`'s workflows are written in `Snakemake`, a popular workflow manager that is highly used in the bioinformatics community. `Snakemake` is highly known for powerful and intuitive workflow design that allows for generating scalable, portable and reproducible workflows. However, `Snakemake`'s declarative naming scheme is very strict. This means that file containing different names will instantly cause the program to fail.

**Below is an example:**
In the `Snakemake`, we need to specify an input, output.

```yaml
rule read_inputs:
    inputs: "path/path/metadata"
    output: "path/path/output.txt"
    script: "scripts/read_inputs.py"
```

The example above is what is known as a `rule`. This is the building block for a `Snakemake` workflow. A `rule` specifies a specific step within your workflow. It requires users to add an input, output and an executable (in this example it's `script`) that will generate the output. In a complete workflow, you will see a series of [rules](https://github.com/snakemake-workflows/rna-seq-star-deseq2/blob/master/workflow/rules/diffexp.smk).

In this example, we are specifying the path to the `input` , `output` and the `script`, which will generate the output.

If we look closely, this `rule` will only work if the `input` file name exactly matches.

Therefore, if the `metadata` folder was renamed to `Metadata`, it will automatically fail despite providing the correct path.

This is where the main issues that @jenna-tomkinson  was having. Since the development of the `CytoSnake` was used `cell-health-data`, the expected naming scheme is identical to the `cell-health-data` naming scheme. 

## Implementation Approach

To solve this issue a dynamic pathing system was developed. What this means that we can pre-define the naming of the file before sending it to `Snakemake's` .

Developing `_paths.yaml` is the star of the this implementation because it predefines paths before submitting into `CytoSnake` workflows. This attempts to solve the issue with `Snakemake`'s strict naming. Therefore, this removes the strict declaration of paths that one needs to do in `Snakemake` workflows. 

Below are the contents of the `_paths.yaml`

```json
{
    "project_dir_path": "/home/erikserrano/Development/CytoSnake/testing",
    "project_dir": {
        "metadata": "/home/erikserrano/Development/CytoSnake/testing/metadata",
        "workflows": "/home/erikserrano/Development/CytoSnake/testing/workflows",
        ".cytosnake": "/home/erikserrano/Development/CytoSnake/testing/.cytosnake",
        "data": "/home/erikserrano/Development/CytoSnake/testing/data",
        "configs": "/home/erikserrano/Development/CytoSnake/testing/configs"
    },
    "config_dir": {
        "configuration": "/home/erikserrano/Development/CytoSnake/testing/configs/configuration.yaml",
        "analysis_configs": {
            "dp_aggregator_config": "/home/erikserrano/Development/CytoSnake/testing/configs/analysis_configs/dp_aggregator_config.yaml",
            "dp_data_configs": "/home/erikserrano/Development/CytoSnake/testing/configs/analysis_configs/dp_data_configs.yaml",
            "single_cell_configs": "/home/erikserrano/Development/CytoSnake/testing/configs/analysis_configs/single_cell_configs.yaml",
            "consensus_configs": "/home/erikserrano/Development/CytoSnake/testing/configs/analysis_configs/consensus_configs.yaml",
            "normalize_configs": "/home/erikserrano/Development/CytoSnake/testing/configs/analysis_configs/normalize_configs.yaml",
            "feature_select_configs": "/home/erikserrano/Development/CytoSnake/testing/configs/analysis_configs/feature_select_configs.yaml",
            "annotate_configs": "/home/erikserrano/Development/CytoSnake/testing/configs/analysis_configs/annotate_configs.yaml"
        }
    },
    "workflow_dir": {
        "workflow": {
            "cp_process": "/home/erikserrano/Development/CytoSnake/testing/workflows/workflow/cp_process",
            "dp_process": "/home/erikserrano/Development/CytoSnake/testing/workflows/workflow/dp_process"
        },
        "envs": {
            "cytominer_env": "/home/erikserrano/Development/CytoSnake/testing/workflows/envs/cytominer_env.yaml",
            "dp_process": "/home/erikserrano/Development/CytoSnake/testing/workflows/envs/dp_process.yaml"
        },
        "scripts": {
            "dp_aggregate": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/dp_aggregate.py",
            "normalize": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/normalize.py",
            "build_dp_consensus": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/build_dp_consensus.py",
            "annotate": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/annotate.py",
            "dp_normalize": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/dp_normalize.py",
            "feature_select": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/feature_select.py",
            "dp_build_consensus": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/dp_build_consensus.py",
            "consensus": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/consensus.py",
            "aggregate_cells": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/aggregate_cells.py",
            "merge_logs": "/home/erikserrano/Development/CytoSnake/testing/workflows/scripts/merge_logs.py"
        },
        "rules": {
            "feature_select": "/home/erikserrano/Development/CytoSnake/testing/workflows/rules/feature_select.smk",
            "preprocessing": "/home/erikserrano/Development/CytoSnake/testing/workflows/rules/preprocessing.smk",
            "merge_logs": "/home/erikserrano/Development/CytoSnake/testing/workflows/rules/merge_logs.smk",
            "dp_process": "/home/erikserrano/Development/CytoSnake/testing/workflows/rules/dp_process.smk",
            "common_cp": "/home/erikserrano/Development/CytoSnake/testing/workflows/rules/common_cp.smk"
        }
    }
}
```

However, this creates another layer of complexity, which is implementing helper functions (next PR). The sole purpose of the helper functions is to declare paths dynamically into `Snakemake` workflows. Therefore, users do not have to worry about having a specific naming scheme in order for the workflows to successfully execute. These functions will interact with the `_paths.yaml` file in order to make pathing declaration much more dynamic. 

Below is an example of how the `helper_functions` will interact with the `configs.yaml` and `_paths.yaml`

Helperfunctions

In addition, this also allows to dynamically set names based on via extensions. One of the best practices of generating outputs names with extensions included. Extensions provide an idea what analysis was conducted within the workflows. For example:

```bash
# common skeleton
{file_name}_{ext1}_{ext2}.csv

# example if config = zcore
SQ00123_norm_zscore.csv

# exaple if config=scAN
SQ00123_norm_scAN.csv
```

The flexibility that the pathing implementation allows to declare input/output names automatically. This makes it much easier to declare input and output file names without the user actively renaming the paths declared under the `rules` within the workflow.

### Greg

Mainly focusing on the design of the implementation. Understanding the order of executing involved in creating a `Project Directory`. Mainly you will be focusing on `cytosnake_setup.py` module. Here are some main things that I w

- Is the current design of setting up a current working directory into a `Project Directory` practical
- Are there any potential assumptions that are not accounted in this implementation.
- Is there a specific case where this implementation will fail when setting up the `Project Direcotry`?
- Any parts of the design that may be unclear.
- Spot any impracticalities.

### Dave

Implementation focus. See how the implemented functions involved in the pathing follow best software development standard: 

Mainly focus on: `cyto_paths.py` `file_utils.py` typically the functions that are called in the `cytosnake_setup.py`

Here are some expectations for your part: 

- Try not to focus too much on the formatting of the code. The next PR will be me manually executing `pre-commits` to format the whole code base.
- pathing pitfalls (main focus) → where the pathing may fail in specific cases. For example creating paths that do not exists due to a user being at a different directory.
- Documentation: see if it reflects the actual function.
- Potential data structure pitfalls. For example: It is better to use `lists` rather than `dict`

## Usage and assumption

CLI

### Init mode

The `init` mode allows users to prepare the current directory into a `project directory`. The `init` mode expect users to provide metadata folder and plate datasets, if done in replicates, then a barcode file must be added as well. 

An example command of using the `init` mode is:

```bash
cytosnake init -d plate_data1.sqlite plate_data2.sqlite -m metadata_dir -b barcodes.txt
```

One can also use wildcards to declare multiple files as well. 

```bash
cytosnake init -d *.sqlite -m metadata_dir -b barcodes.txt
```

Once a user inputs the required files, the current directory gets transformed into a `Project Drectory`. A `Project Directory` allows `CytoSnake` to know that the files used to initialize in the current directory is being prepared for analysis. `CytoSnake` uses the `.cytosnake` directory as a landmark to know that a project is being conducting in the current directory. (similar to how git recognizes a directory as a repo by using `.git`) 

The `.cytosnake` folder has two purposes: 

1. To let `CytoSnake` know that the directory is a project directory
2. contains a `_paths.yaml` file that provides `CytoSnake` pathing information.

There is more happening in the background when converting a current directory to a `Project Directory`. Assuming that `CytoSnake` has been `pip` installed, the `init` function makes a request to transfer the necessary files in order to conduct any analysis. 

If you look at the image above where it says `CytoSnake Package` , we see that `init` mode makes a call in order to load in the `configs` and the `workflow` folders. 

### Run mode

Run mode allows `CytoSnake` to execute workflows found within the `workflows/` directory folder. 

Since `CytoSnake` already knows what the inputs are present due to the `_paths.yaml` folder, users all need to do is type:

```bash
# executing the cp_process workflow
cytosnake run cp_process

# executing cp_rpocess worflow with 9 cores
cytosnake run cp_process -c 9
```

### Help mode

The help mode is executed by typing:

```bash
cytosnake help
```

This will print out the whole CLI documentation with the three modes together. 

If you are only interested reading documentation of one mode, you can simply type:

```bash
cytosnake run help
```

This will only print out the `help` documentation for only the `run` mode

### CODE REVIEW

File: cytosnake/guards/path_guards.py
Comment (gwaybio) [inline cytosnake/guards/path_guards.py:6]:
```suggestion
Functions for checking paths. 
```

Does this still capture your meaning? I don't want to simplify too much!

File: cytosnake/helpers/helper_funcs.py
Comment (gwaybio) [inline cytosnake/helpers/helper_funcs.py:4]:
```suggestion
This module contains helper functions for snakemake workflows. 
```

Please edit my suggestion if it removes your meaning

File: cytosnake/helpers/helper_funcs.py
Comment (gwaybio) [inline cytosnake/helpers/helper_funcs.py:30]:
```suggestion
        raise KeyError("Unable to find data folder path in project_dir config")
```

True?

File: cytosnake/helpers/helper_funcs.py
Comment (gwaybio) [inline cytosnake/helpers/helper_funcs.py:35]:
```suggestion
    """Returns absolute path were the data are located
```

the word "data" is plural :)

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [inline cytosnake/utils/cyto_paths.py:33]: Is this PR out of scope to perform this TODO?

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [inline cytosnake/utils/cyto_paths.py:83]: this function seems to require cytosnake to run in a git configured directory. Is this necessary? I think we will always use cytosnake in a github repo, but I don't think it will be universally true. Am I not thinking of the reason why `.git` is necessary?

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [inline cytosnake/utils/cyto_paths.py:91]:
```suggestion
    """Returns complete path where cytosnake performs the analysis. The function
```

Active voice

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [inline cytosnake/utils/cyto_paths.py:106]:
```suggestion
    project_dir = Path().absolute()
```

major bug still, or safe to delete?

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [inline cytosnake/utils/cyto_paths.py:116]: what is an `fpath`? Is the `f` a typo?

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [inline cytosnake/utils/cyto_paths.py:151]:
```suggestion
        raise FileNotFoundError("Unable to find config directory")
```

I think?

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [inline cytosnake/utils/cyto_paths.py:174]: I get itchy seeing this hardcoded! 

I can see a real danger in this not working if someone has a different unwanted directory. Maybe specifying a specific file to find in the glob? (that is also dangerous!) not sure how to handle!

File: cytosnake/utils/cytosnake_setup.py
Comment (gwaybio) [inline cytosnake/utils/cytosnake_setup.py:47]: This pattern seems to be repeated a bit - I wonder if you can make a generic function to perform. (Discussion point!)

File: cytosnake/utils/cytosnake_setup.py
Comment (gwaybio) [inline cytosnake/utils/cytosnake_setup.py:50]: Is this PR out of scope to perform this TODO?

**Review by gwaybio — CHANGES_REQUESTED** [review]
Mostly minor points but some fairly important discussion items. I'm looking forward to your comments and for @d33bs review!

**Review by d33bs — COMMENTED** [review]
Great work! I left a few comments and suggestions. Using your PR description as a reference, I wanted to provide additional comments here: 

> - Try not to focus too much on the formatting of the code. The next PR will be me manually executing pre-commits to format the whole code base.

When you can, I recommend using pre-commit checks / fixes as you develop and before a PR is issued. This can help save you time from receiving comments about content that otherwise may be fixed ahead of time. Without the changes themselves it may be difficult for reviewers to understand what to ignore or not (pre-commit checks can do many things).

> - pathing pitfalls (main focus) → where the pathing may fail in specific cases. For example creating paths that do not exists due to a user being at a different directory.

I focused on looking at this and providing thoughts were possible. I also felt that it could be beneficial to provide a programmatic test-case so you may quickly validate expected outcomes from what the new functionality is providing. Doing this would follow SnakeMake's best practices: _"When publishing your workflow in a Github repository, it is a good idea to add some minimal test data and configure Github Actions for continuously testing the workflow on each new commit."_ ([reference](https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html)) This could potentially serve multiple needs: it shares results during development, demonstrates to reviewers how the code should operate during implementation, and also may inform users as they implement functionality with the package. Testing could be as minimal as one test to cover all functionality or many more modular ones to keep the complexity low for development.

> - Documentation: see if it reflects the actual function.

I left comments and suggestions where I found items that might need to be addressed. Generally here I have the same recommendation as above to write tests which can serve as functional documentation of expected functionality. Code-based examples within docstrings can also serve as a way to demonstrate intended use, for ex [see here](https://stackoverflow.com/a/64452749). Also, just in case you meant the docs under `docs/`, I didn't notice there were changes there in this PR - but if I should take a look there too please don't hesitate to let me know.

> - Potential data structure pitfalls. For example: It is better to use lists rather than dict

I think a few of my comments related to this topic and generally don't have too much more to add at this time. While it didn't pertain to this PR, I noticed the use of `random` in `generate_sample_inputs.py`. It may be helpful to consider the use of [`faker`](https://github.com/joke2k/faker) with regards to the earlier mention above on testing.

File: cytosnake/common/errors.py
Comment (axiomcura) [reply cytosnake/common/errors.py:120]: In my editor, all the different raises appear together. I do not know if this is just a `vscode` only problem or not.

File: cytosnake/cli/cmd.py
Comment (axiomcura) [reply cytosnake/cli/cmd.py:105]: as of right now there is no Logging system implemented yet. This will be noted here! [#9](https://github.com/WayScience/CytoSnake/issues/9)

File: cytosnake/cli/cmd.py
Comment (axiomcura) [reply cytosnake/cli/cmd.py:112]: The `WorkflowSearchPath` will only raise an exception if 

These `cases` focus on mode selection. If the user provides an invalid `mode` name then the `case _` is executed.

File: cytosnake/cli/args.py
Comment (axiomcura) [reply cytosnake/cli/args.py:69]: @gwaybio I am not quiet understanding what you mean. This class check if the workflow that the user inputs exists. This is done by loading in the `_paths.yaml` file, and search to see if the workflow name exists within the file.

Comment (gwaybio) [reply cytosnake/cli/args.py:69]: the code as currently written assigns the output of `load_workflow_path(values)` to a variable named `workflow_paths`. Line 69 then sets this variable named `workflow_paths` to a new variable called `values`. 

My suggestion here was to simply skip the middle path and simply set `load_workflow_path(values)` to `values`

(unless my understanding of this function is incorrect, then I think my suggestion is cleaner)

Comment (gwaybio) [reply cytosnake/cli/args.py:69]: the code as currently written assigns the output of `load_workflow_path(values)` to a variable named `workflow_paths`. Line 69 then sets this variable named `workflow_paths` to a new variable called `values`. 

My suggestion here was to simply skip the middle path and simply set `load_workflow_path(values)` to `values`

(unless my understanding of this function is incorrect, then I think my suggestion is cleaner)

File: cytosnake/common/errors.py
Comment (axiomcura) [reply cytosnake/common/errors.py:153]: From what I learned from my Software Engineering class, python exceptions do no raise non-zero exceptions if the exception trigger is true. Therefore, the reason why I added a `sys.exit` with an exit code is to ensure a non-zero exit code is always returned when exceptions are raised. 

Hopefully this makes sense.

File: cytosnake/common/errors.py
Comment (axiomcura) [reply cytosnake/common/errors.py:92]: The purpose of `display_error` is to ensure that the CLI tool exits non-zero exit status. It also allows to edit exception messages if needed in order to make error messages more understandable. So in the future, if CytoSnake contains it's own error code, the display error will be extremely helpful. 

How do you see `Logger.error()` play a role with `display_error`

File: cytosnake/common/errors.py
Comment (axiomcura) [reply cytosnake/common/errors.py:84]: `--force` This is a good point to consider. Unfortunately, `--force` is a flag for `Snakemake`'s API. For example, if you interupt the workflow process, a `lock` is added into the workflow file. Therefore, if you try to run it again, it will will raise an error. `--force` allows the user to bypass that lock that `Snakemake` annoyingly adds.

File: cytosnake/utils/cytosnake_setup.py
Comment (axiomcura) [reply cytosnake/utils/cytosnake_setup.py:50]: Yeah I was getting some pesky circular import errors and I have not figured that out yet. Might be a seperate PR. This is a temporary solution.

File: cytosnake/utils/cytosnake_setup.py
Comment (axiomcura) [reply cytosnake/utils/cytosnake_setup.py:47]: That's currently on the works. Probably might see this change in the `helper_functions` PR

File: cytosnake/utils/cyto_paths.py
Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:116]: fpaths indicates that there are multiple files paths being returned

File: cytosnake/utils/cyto_paths.py
Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:83]: I am using `.git` as a landmark to find where the project folder is installed. I guess it's a hacky way to identify where the package is installed in your environment. if there is a more practical way in how to do this I am all ears :)

File: cytosnake/common/errors.py
Comment (d33bs) [reply cytosnake/common/errors.py:92]: I'd recommend looking into how [`logging`](https://docs.python.org/3/library/logging.html) might be used with some of this work in order to take advantage of built-in capabilities. Doing this may standardize how log messages are displayed or stored. I bring this up as a suggestion towards increasing development velocity.

File: cytosnake/cli/args.py
Comment (axiomcura) [reply cytosnake/cli/args.py:69]: I see what you mean! got it!

File: cytosnake/common/errors.py
Comment (axiomcura) [reply cytosnake/common/errors.py:92]: Will definitely take you up on that! Thanks!

File: cytosnake/utils/cyto_paths.py
Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:33]: Yes this is currently out of the scope. I let this TODO message for the `helper_functions` PR

File: cytosnake/utils/cytosnake_setup.py
Comment (axiomcura) [reply cytosnake/utils/cytosnake_setup.py:94]: Yes it would make sense. However, lots of other functions use this call as well. Might be a seperate PR in order to redesign the function.

**Comment by axiomcura** [comment]
Hello @d33bs and @gwaybio 

Hopefully I have covered all your comments! Ready for the next round!

Thank you for all your constructive comments!

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [reply cytosnake/utils/cyto_paths.py:83]: Yeah, it does seem hacky. Maybe ok for now?

As I see it, cytosnake will not work if used outside of a github repo. I think we want to support non-git usage.

Could `init` have some sort of `--project_folder` command argument that defaults to `pwd` (or `.`)?

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [reply cytosnake/utils/cyto_paths.py:171]: Gotcha! What is your plan then? (ignore for now is an option)

File: cytosnake/utils/cytosnake_setup.py
Comment (gwaybio) [reply cytosnake/utils/cytosnake_setup.py:47]: This sounds like a good option to create an issue?

File: cytosnake/utils/cytosnake_setup.py
Comment (gwaybio) [reply cytosnake/utils/cytosnake_setup.py:50]: Issue?

**Review by gwaybio — APPROVED** [review]
There are a couple comments still unresolved, but good to merge once they are addressed! Nice work! (love the diagrams! I wonder if you can preserve them (and the text!) somewhere more visible 🤔 )

File: cytosnake/utils/cytosnake_setup.py
Comment (d33bs) [reply cytosnake/utils/cytosnake_setup.py:94]: Sounds good, I understand.

File: cytosnake/utils/file_utils.py
Comment (d33bs) [inline cytosnake/utils/file_utils.py:74]:
```suggestion
        if the path is unable to be found
```

**Review by d33bs — CHANGES_REQUESTED** [review]
Nice revisions! I made a few comments, one of which should be addressed before merging (on the use of `logging`). Please don't hesitate to let me know if you have any questions.

File: cytosnake/utils/cyto_paths.py
Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:116]: I created an issue that will be taken care of in the `helper_fuction`PR #27

File: cytosnake/utils/cyto_paths.py
Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:171]: First i need to get a better idea how the file structure is going to be once I make `CytoSnake` pip installable. If we need a quick fix, then putting these files names in a config file will also work.

File: cytosnake/utils/cytosnake_setup.py
Comment (axiomcura) [reply cytosnake/utils/cytosnake_setup.py:50]: I agree. Created issue #28

File: cytosnake/utils/cytosnake_setup.py
Comment (axiomcura) [reply cytosnake/utils/cytosnake_setup.py:47]: Created issue #29. Will add this to my checklist when developing `helper_fuction` PR

File: cytosnake/utils/cyto_paths.py
Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:83]: I think there is some confusion. The `get_cytosnake_package_path()` returns the absolute path of `CytoSnake` within the python env. For example, after pip installing, we want the path where `CytoSnake` is installed in the python env.

Comment (gwaybio) [reply cytosnake/utils/cyto_paths.py:83]: but returning the absolute path of Cytosnake requires there to be a `.git` folder?

Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:83]: Yeah. I have not figure out a way on how to get the package path without using `.git` folder. I was doing some reading in using the `__name__` can help resolve this.

File: cytosnake/cli/cmd.py
Comment (axiomcura) [reply cytosnake/cli/cmd.py:58]: Hmm this is actually a good idea for [#9](https://github.com/WayScience/CytoSnake/issues/9). I'll keep this in mind. One of the issues I have trouble understanding is how `Snakemake` handles logs and see how much control developers (me and contributors) have. 

Based of what you have suggested, it seems like implementing a logging system using the `logging` module may be an option if the `Snakemake` native logging system does not satisfy our needs.

File: cytosnake/common/errors.py
Comment (axiomcura) [reply cytosnake/common/errors.py:153]: I feel once I have the logging implementation complete [#9](https://github.com/WayScience/CytoSnake/issues/9), this code base may be changed or removed.

**Comment by axiomcura** [comment]
@d33bs @gwaybio 

Hopefully I have answered all your comments! Some comments brought interesting points that were added into the issues section.  

 I'd be happy to merge this PR right now If there aren't any objections.

**Comment by gwaybio** [comment]
you already have my approval 😄 - feel free to merge when _you_ are happy with it.

**Comment by axiomcura** [comment]
Sounds good! Merging!

**Pull request merged**
